lib/system/assertions.nim

```diff
@@ -82,29 +82,40 @@ template onFailedAssert*(msg, code: untyped): untyped {.dirty.} =
 
 template doAssertRaises*(exception: typedesc, code: untyped) =
   ## Raises ``AssertionDefect`` if specified ``code`` does not raise the
-  ## specified exception. Example:
+  ## specified exception. Use `exception = void` to specify a foreign exception.
```

Comment:
If the caller needs to know this and passes a special value to this, why not instead create a ``doAssertRaisesForeign``. Or add nothing, you might as well write a ``try .. except`` block. Or make ``doAssertRaises`` smarter so that it detects "foreign exceptions" on its own.

Reply:
fixed by adding a catch-all `doAssertRaises` overload that catches every raise including foreign exceptions.

Reply:
I don't understand this solution. I didn't ask for new overloads, I asked for a detection of foreign exceptions and your code seems to accomplish that, what do you need the separate overload for?

Reply:
> what do you need the separate overload for?

the PR as written makes both those programs correct:

* P1: this will correctly raise `AssertionDefect` (and give proper message) since code excepts ValueError but a foreign exception was raised
```nim
when true:
  {.emit:"""
  #include <stdexcept>
  void fn(){throw std::runtime_error("asdf");}""".}
  proc fn(){.importcpp.}
  doAssertRaises(ValueError): fn()
```

* P2: this will correctly succeed (ie, not raise `AssertionDefect`) since `fn` raises something (a foreign exception)
```nim
when true:
  {.emit:"""
  #include <stdexcept>
  void fn(){throw std::runtime_error("asdf");}""".}
  proc fn(){.importcpp.}
  doAssertRaises: fn()
```

if I removed the overload, P2 wouldn't be possible to write

Reply:
This overload overloads ``untyped`` on an inconsistent position so it's unlikely to work well. But more importantly nobody asked for this feature addition in a bugfix PR.

Reply:
> feature addition in a bugfix PR.

I've removed it from this PR